### DISCUSSION TITLE
Can SliderInt, DragFloatRange, etc. work with mutable struct fields?

### DISCUSSION THREAD

**Discussion by claforte** [Gnimuc/CImGui.jl#49] [General]
Hi Gnimuc, thanks again for this fantastic module! I'm starting to get the hang of it and I haven't had that much fun designing a UI in years!

I'm now trying to use `SliderInt` and `DragFloatRange2` with the fields of mutable structs. Is that possible? e.g. something like:

```
mutable struct ImgView
    img::Union{Img, Missing} # If missing, this view is empty. The texture_id should be -1 as well in that case.
    texture_id::Int64 # GL texture ID. If -1, this is view is dirty, needs to be reprocessed.

    black::Float16 # min value for brightness adjustment
    white::Float16 # max value for brightness adjustment

    # Parameters that influence the Img loaded:
    epoch::Int64 # epoch number. 0-based.
    batch::Int64 # batch number within the epoch. 0-based.
    item::Int64 # item number within the batch. 0-based.
    ds::String # dataset, can be "train" or "valid"
end

...

function show_params(v::ImgView)
    @cstatic epoch=Cint(0) batch=Cint(0) black=Cfloat(0.0) white=Cfloat(1.0) begin
        @c CImGui.SliderInt("Epoch", &(v.epoch), 0, 3)   # THIS FAILS with MethodError
        @c CImGui.DragFloatRange2("Intensity", &black, &white, 0.1, -100.0, 100.0, "Black: %.1f", "White: %.1f")
    end
end
```

The above works when I strictly use @cstatic variables, but it doesn't work with fields of my mutable structs. I want to have different values/range for each instance of the ImgView. Is there a trick to do this, or can you recommend a work-around? Thanks!

**Comment by Gnimuc** [comment]
`epoch=Cint(0) batch=Cint(0) black=Cfloat(0.0) white=Cfloat(1.0)` 

Have you tried to use these correct types in your `ImgView` struct? The `@c` macro will create a `Ref(v.epoch)`, but the type is mismatched, so the code may yield `MethodError`s.

**Reply by claforte** [reply]
Thanks Gnimuc for your super fast response! That did the trick!
